File: sdk/csharp/RaptoreumSdk/RaptoreumTransaction.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Security.Cryptography;

namespace RaptoreumSdk
{
    public class TxInput
    {
        public string Txid { get; set; } = "";
        public uint Vout { get; set; }
        public byte[] ScriptPubKey { get; set; } = Array.Empty<byte>();
        public ulong Amount { get; set; }
        public byte[] ScriptSig { get; set; } = Array.Empty<byte>();
    }

    public class TxOutput
    {
        public ulong Value { get; set; }
        public byte[] Script { get; set; } = Array.Empty<byte>();
    }

    public class UTXO
    {
        public string txid { get; set; } = "";
        public uint vout { get; set; }
        public double amount { get; set; }
        public string scriptPubKey { get; set; } = "";
    }

    public class RaptoreumTransactionBuilder
    {
        public List<TxInput> Inputs { get; } = new List<TxInput>();
        public List<TxOutput> Outputs { get; } = new List<TxOutput>();
        public uint Locktime { get; set; } = 0;
        public uint Version { get; set; } = 1;

        public void AddInput(string txid, uint vout, string scriptPubKey, double amountRtm)
        {
            Inputs.Add(new TxInput
            {
                Txid = txid,
                Vout = vout,
                ScriptPubKey = HexDecode(scriptPubKey),
                Amount = (ulong)Math.Round(amountRtm * 100000000)
            });
        }

        public void AddOutput(string address, double amountRtm)
        {
            byte[] hash160 = AddressToHash160(address);
            // P2PKH scriptPubKey: OP_DUP OP_HASH160 <hash160> OP_EQUALVERIFY OP_CHECKSIG
            byte[] script = new byte[25];
            script[0] = 0x76;
            script[1] = 0xa9;
            script[2] = 0x14;
            Array.Copy(hash160, 0, script, 3, 20);
            script[23] = 0x88;
            script[24] = 0xac;

            Outputs.Add(new TxOutput
            {
                Value = (ulong)Math.Round(amountRtm * 100000000),
                Script = script
            });
        }

        public byte[] Serialize()
        {
            using var ms = new MemoryStream();
            // Version
            byte[] verBytes = BitConverter.GetBytes(Version);
            if (!BitConverter.IsLittleEndian) Array.Reverse(verBytes);
            ms.Write(verBytes, 0, 4);

            // Inputs
            byte[] inCount = EncodeVarInt(Inputs.Count);
            ms.Write(inCount, 0, inCount.Length);

            foreach (var input in Inputs)
            {
                byte[] txidBytes = HexDecode(input.Txid);
                Array.Reverse(txidBytes); // Little endian
                ms.Write(txidBytes, 0, txidBytes.Length);

                byte[] voutBytes = BitConverter.GetBytes(input.Vout);
                if (!BitConverter.IsLittleEndian) Array.Reverse(voutBytes);
                ms.Write(voutBytes, 0, 4);

                byte[] scriptSigLen = EncodeVarInt(input.ScriptSig.Length);
                ms.Write(scriptSigLen, 0, scriptSigLen.Length);
                ms.Write(input.ScriptSig, 0, input.ScriptSig.Length);

                byte[] seqBytes = BitConverter.GetBytes(0xffffffff);
                ms.Write(seqBytes, 0, 4);
            }

            // Outputs
            byte[] outCount = EncodeVarInt(Outputs.Count);
            ms.Write(outCount, 0, outCount.Length);

            foreach (var output in Outputs)
            {
                byte[] valBytes = BitConverter.GetBytes(output.Value);
                if (!BitConverter.IsLittleEndian) Array.Reverse(valBytes);
                ms.Write(valBytes, 0, 8);

                byte[] scriptLen = EncodeVarInt(output.Script.Length);
                ms.Write(scriptLen, 0, scriptLen.Length);
                ms.Write(output.Script, 0, output.Script.Length);
            }

            // Locktime
            byte[] ltBytes = BitConverter.GetBytes(Locktime);
            if (!BitConverter.IsLittleEndian) Array.Reverse(ltBytes);
            ms.Write(ltBytes, 0, 4);

            return ms.ToArray();
        }

        public void Sign(byte[] privateKeyBytes)
        {
            // Derive compressed public key
            using var ecdsa = ECDsa.Create(new ECParameters
            {
                Curve = ECCurve.CreateFromFriendlyName("secp256k1"),
                D = privateKeyBytes
            });
            var parameters = ecdsa.ExportParameters(false);
            var qX = parameters.Q.X ?? throw new InvalidOperationException("Invalid X");
            var qY = parameters.Q.Y ?? throw new InvalidOperationException("Invalid Y");

            byte prefix = (qY[qY.Length - 1] % 2 == 0) ? (byte)0x02 : (byte)0x03;
            byte[] pubBytes = new byte[33];
            pubBytes[0] = prefix;
            Array.Copy(qX, 0, pubBytes, 1, 32);

            for (int i = 0; i < Inputs.Count; i++)
            {
                byte[][] originalScriptSigs = new byte[Inputs.Count][];
                for (int j = 0; j < Inputs.Count; j++)
                {
                    originalScriptSigs[j] = Inputs[j].ScriptSig;
                }

                for (int j = 0; j < Inputs.Count; j++)
                {
                    if (j == i)
                    {
                        Inputs[j].ScriptSig = Inputs[j].ScriptPubKey;
                    }
                    else
                    {
                        Inputs[j].ScriptSig = Array.Empty<byte>();
                    }
                }

                byte[] serialized = Serialize();
                byte[] preimage = new byte[serialized.Length + 4];
                Array.Copy(serialized, preimage, serialized.Length);
                preimage[serialized.Length] = 0x01; // SIGHASH_ALL

                byte[] sig = RaptoreumWallet.SignMessage(privateKeyBytes, preimage);
                byte[] sigWithHash = new byte[sig.Length + 1];
                Array.Copy(sig, sigWithHash, sig.Length);
                sigWithHash[sig.Length] = 0x01; // SIGHASH_ALL

                byte[] scriptSig = new byte[1 + sigWithHash.Length + 1 + pubBytes.Length];
                scriptSig[0] = (byte)sigWithHash.Length;
                Array.Copy(sigWithHash, 0, scriptSig, 1, sigWithHash.Length);
                scriptSig[1 + sigWithHash.Length] = (byte)pubBytes.Length;
                Array.Copy(pubBytes, 0, scriptSig, 1 + sigWithHash.Length + 1, pubBytes.Length);

                for (int j = 0; j < Inputs.Count; j++)
                {
                    Inputs[j].ScriptSig = originalScriptSigs[j];
                }
                Inputs[i].ScriptSig = scriptSig;
            }
        }

        public static (List<UTXO> selected, double fee) SelectInputs(List<UTXO> utxos, double targetAmountRtm, uint feeRateSatByte = 1)
        {
            ulong targetSat = (ulong)Math.Round(targetAmountRtm * 100000000);
            ulong accumulated = 0;
            List<UTXO> selected = new List<UTXO>();
            uint numOutputs = 2;

            foreach (var utxo in utxos)
            {
                selected.Add(utxo);
                accumulated += (ulong)Math.Round(utxo.amount * 100000000);

                uint size = 148 * (uint)selected.Count + 34 * numOutputs + 10;
                ulong fee = size * feeRateSatByte;

                if (accumulated >= targetSat + fee)
                {
                    return (selected, (double)fee);
                }
            }

            throw new InvalidOperationException("Insufficient funds");
        }

        private static byte[] EncodeVarInt(long n)
        {
            if (n < 0xfd)
            {
                return new byte[] { (byte)n };
            }
            else if (n <= 0xffff)
            {
                byte[] buf = new byte[3];
                buf[0] = 0xfd;
                byte[] temp = BitConverter.GetBytes((ushort)n);
                if (!BitConverter.IsLittleEndian) Array.Reverse(temp);
                Array.Copy(temp, 0, buf, 1, 2);
                return buf;
            }
            else if (n <= 0xffffffff)
            {
                byte[] buf = new byte[5];
                buf[0] = 0xfe;
                byte[] temp = BitConverter.GetBytes((uint)n);
                if (!BitConverter.IsLittleEndian) Array.Reverse(temp);
                Array.Copy(temp, 0, buf, 1, 4);
                return buf;
            }
            else
            {
                byte[] buf = new byte[9];
                buf[0] = 0xff;
                byte[] temp = BitConverter.GetBytes((ulong)n);
                if (!BitConverter.IsLittleEndian) Array.Reverse(temp);
                Array.Copy(temp, 0, buf, 1, 8);
                return buf;
            }
        }

        private static byte[] AddressToHash160(string address)
        {
            const string B58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
            System.Numerics.BigInteger n = 0;
            for (int i = 0; i < address.Length; i++)
            {
                int idx = B58.indexOf_custom(address[i]); // wait! custom index helper since .NET Standard 2.0 indexof can be case-sensitive or different
                if (idx == -1) throw new FormatException("Invalid Base58 char");
                n = n * 58 + idx;
            }
            byte[] bytes = n.ToByteArray();
            Array.Reverse(bytes);
            if (bytes.Length > 25 && bytes[0] == 0)
            {
                byte[] temp = new byte[bytes.Length - 1];
                Array.Copy(bytes, 1, temp, 0, temp.Length);
                bytes = temp;
            }
            if (bytes.Length < 25)
            {
                byte[] padded = new byte[25];
                Array.Copy(bytes, 0, padded, 25 - bytes.Length, bytes.Length);
                bytes = padded;
            }
            byte[] payload = new byte[21];
            byte[] checksum = new byte[4];
            Array.Copy(bytes, 0, payload, 0, 21);
            Array.Copy(bytes, 21, checksum, 0, 4);

            using var sha = SHA256.Create();
            byte[] h1 = sha.ComputeHash(payload);
            byte[] h2 = sha.ComputeHash(h1);

            for (int i = 0; i < 4; i++)
            {
                if (h2[i] != checksum[i]) throw new FormatException("Invalid address checksum");
            }

            byte[] hash160 = new byte[20];
            Array.Copy(payload, 1, hash160, 0, 20);
            return hash160;
        }

        private static byte[] HexDecode(string hex)
        {
            byte[] bytes = new byte[hex.Length / 2];
            for (int i = 0; i < bytes.Length; i++)
            {
                bytes[i] = Convert.ToByte(hex.Substring(i * 2, 2), 16);
            }
            return bytes;
        }
    }

    internal static class StringExtensions
    {
        public static int indexOf_custom(this string s, char c)
        {
            for (int i = 0; i < s.Length; i++)
            {
                if (s[i] == c) return i;
            }
            return -1;
        }
    }
}

File: sdk/fsharp/RaptoreumSdk/RaptoreumTransaction.fs
namespace RaptoreumSdk

open System
open System.Collections.Generic
open System.IO
open System.Security.Cryptography

type TxInput() =
    member val Txid = "" with get, set
    member val Vout = 0u with get, set
    member val ScriptPubKey = Array.empty<byte> with get, set
    member val Amount = 0UL with get, set
    member val ScriptSig = Array.empty<byte> with get, set

type TxOutput() =
    member val Value = 0UL with get, set
    member val Script = Array.empty<byte> with get, set

type UTXO() =
    member val txid = "" with get, set
    member val vout = 0u with get, set
    member val amount = 0.0 with get, set
    member val scriptPubKey = "" with get, set

type RaptoreumTransactionBuilder() =
    let inputs = new List<TxInput>()
    let outputs = new List<TxOutput>()
    
    member val Locktime = 0u with get, set
    member val Version = 1u with get, set

    member this.Inputs = inputs
    member this.Outputs = outputs

    static member private HexDecode(hex: string) : byte[] =
        let bytes = Array.zeroCreate (hex.Length / 2)
        for i in 0 .. bytes.Length - 1 do
            bytes.[i] <- Convert.ToByte(hex.Substring(i * 2, 2), 16)
        bytes

    static member private EncodeVarInt(n: int64) : byte[] =
        if n < 0xfdL then
            [| byte n |]
        elif n <= 0xffffL then
            let buf = Array.zeroCreate 3
            buf.[0] <- 0xfduy
            let temp = BitConverter.GetBytes(uint16 n)
            if not BitConverter.IsLittleEndian then Array.Reverse(temp)
            Array.Copy(temp, 0, buf, 1, 2)
            buf
        elif n <= 0xffffffffL then
            let buf = Array.zeroCreate 5
            buf.[0] <- 0xfeuy
            let temp = BitConverter.GetBytes(uint32 n)
            if not BitConverter.IsLittleEndian then Array.Reverse(temp)
            Array.Copy(temp, 0, buf, 1, 4)
            buf
        else
            let buf = Array.zeroCreate 9
            buf.[0] <- 0xffuy
            let temp = BitConverter.GetBytes(uint64 n)
            if not BitConverter.IsLittleEndian then Array.Reverse(temp)
            Array.Copy(temp, 0, buf, 1, 8)
            buf

    static member private AddressToHash160(address: string) : byte[] =
        let B58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
        let mutable n = System.Numerics.BigInteger.Zero
        for i in 0 .. address.Length - 1 do
            let idx = B58.IndexOf(address.[i])
            if idx = -1 then raise (new FormatException("Invalid Base58 char"))
            n <- n * new System.Numerics.BigInteger(58) + new System.Numerics.BigInteger(idx)
            
        let mutable bytes = n.ToByteArray()
        Array.Reverse(bytes)
        if bytes.Length > 25 && bytes.[0] = 0uy then
            let temp = Array.zeroCreate (bytes.Length - 1)
            Array.Copy(bytes, 1, temp, 0, temp.Length)
            bytes <- temp
        if bytes.Length < 25 then
            let padded = Array.zeroCreate 25
            Array.Copy(bytes, 0, padded, 25 - bytes.Length, bytes.Length)
            bytes <- padded
            
        let payload = Array.zeroCreate 21
        let checksum = Array.zeroCreate 4
        Array.Copy(bytes, 0, payload, 0, 21)
        Array.Copy(bytes, 21, checksum, 0, 4)

        using (SHA256.Create()) (fun sha ->
            let h1 = sha.ComputeHash(payload)
            let h2 = sha.ComputeHash(h1)
            for i in 0 .. 3 do
                if h2.[i] <> checksum.[i] then raise (new FormatException("Invalid address checksum"))
        )

        let hash160 = Array.zeroCreate 20
        Array.Copy(payload, 1, hash160, 0, 20)
        hash160

    member this.AddInput(txid: string, vout: uint32, scriptPubKey: string, amountRtm: double) =
        let inVal = new TxInput()
        inVal.Txid <- txid
        inVal.Vout <- vout
        inVal.ScriptPubKey <- RaptoreumTransactionBuilder.HexDecode(scriptPubKey)
        inVal.Amount <- uint64 (Math.Round(amountRtm * 100000000.0))
        inputs.Add(inVal)

    member this.AddOutput(address: string, amountRtm: double) =
        let hash160 = RaptoreumTransactionBuilder.AddressToHash160(address)
        let script = Array.zeroCreate 25
        script.[0] <- 0x76uy
        script.[1] <- 0xa9uy
        script.[2] <- 0x14uy
        Array.Copy(hash160, 0, script, 3, 20)
        script.[23] <- 0x88uy
        script.[24] <- 0xacuy
        
        let outVal = new TxOutput()
        outVal.Value <- uint64 (Math.Round(amountRtm * 100000000.0))
        outVal.Script <- script
        outputs.Add(outVal)

    member this.Serialize() : byte[] =
        using (new MemoryStream()) (fun ms ->
            let verBytes = BitConverter.GetBytes(this.Version)
            if not BitConverter.IsLittleEndian then Array.Reverse(verBytes)
            ms.Write(verBytes, 0, 4)

            let inCount = RaptoreumTransactionBuilder.EncodeVarInt(int64 inputs.Count)
            ms.Write(inCount, 0, inCount.Length)

            for input in inputs do
                let txidBytes = RaptoreumTransactionBuilder.HexDecode(input.Txid)
                Array.Reverse(txidBytes)
                ms.Write(txidBytes, 0, txidBytes.Length)

                let voutBytes = BitConverter.GetBytes(input.Vout)
                if not BitConverter.IsLittleEndian then Array.Reverse(voutBytes)
                ms.Write(voutBytes, 0, 4)

                let scriptSigLen = RaptoreumTransactionBuilder.EncodeVarInt(int64 input.ScriptSig.Length)
                ms.Write(scriptSigLen, 0, scriptSigLen.Length)
                ms.Write(input.ScriptSig, 0, input.ScriptSig.Length)

                let seqBytes = BitConverter.GetBytes(0xffffffffu)
                ms.Write(seqBytes, 0, 4)

            let outCount = RaptoreumTransactionBuilder.EncodeVarInt(int64 outputs.Count)
            ms.Write(outCount, 0, outCount.Length)

            for output in outputs do
                let valBytes = BitConverter.GetBytes(output.Value)
                if not BitConverter.IsLittleEndian then Array.Reverse(valBytes)
                ms.Write(valBytes, 0, 8)

                let scriptLen = RaptoreumTransactionBuilder.EncodeVarInt(int64 output.Script.Length)
                ms.Write(scriptLen, 0, scriptLen.Length)
                ms.Write(output.Script, 0, output.Script.Length)

            let ltBytes = BitConverter.GetBytes(this.Locktime)
            if not BitConverter.IsLittleEndian then Array.Reverse(ltBytes)
            ms.Write(ltBytes, 0, 4)

            ms.ToArray()
        )

    member this.Sign(privateKeyBytes: byte[]) =
        let mutable ecParams = new ECParameters()
        ecParams.Curve <- ECCurve.CreateFromFriendlyName("secp256k1")
        ecParams.D <- privateKeyBytes
        using (ECDsa.Create(ecParams)) (fun ecdsa ->
            let parameters = ecdsa.ExportParameters(false)
            let qX = parameters.Q.X
            let qY = parameters.Q.Y
            if qX = null || qY = null then raise (new InvalidOperationException("Invalid key points"))
            
            let prefix = if qY.[qY.Length - 1] % 2uy = 0uy then 0x02uy else 0x03uy
            let pubBytes = Array.zeroCreate 33
            pubBytes.[0] <- prefix
            Array.Copy(qX, 0, pubBytes, 1, 32)

            for i in 0 .. inputs.Count - 1 do
                let originalScriptSigs = Array.zeroCreate inputs.Count
                for j in 0 .. inputs.Count - 1 do
                    originalScriptSigs.[j] <- inputs.[j].ScriptSig

                for j in 0 .. inputs.Count - 1 do
                    if j = i then
                        inputs.[j].ScriptSig <- inputs.[j].ScriptPubKey
                    else
                        inputs.[j].ScriptSig <- Array.empty<byte>

                let serialized = this.Serialize()
                let preimage = Array.zeroCreate (serialized.Length + 4)
                Array.Copy(serialized, preimage, serialized.Length)
                preimage.[serialized.Length] <- 0x01uy // SIGHASH_ALL

                let sigVal = RaptoreumWallet.SignMessage(privateKeyBytes, preimage)
                let sigWithHash = Array.zeroCreate (sigVal.Length + 1)
                Array.Copy(sigVal, sigWithHash, sigVal.Length)
                sigWithHash.[sigVal.Length] <- 0x01uy // SIGHASH_ALL

                let scriptSig = Array.zeroCreate (1 + sigWithHash.Length + 1 + pubBytes.Length)
                scriptSig.[0] <- byte sigWithHash.Length
                Array.Copy(sigWithHash, 0, scriptSig, 1, sigWithHash.Length)
                scriptSig.[1 + sigWithHash.Length] <- byte pubBytes.Length
                Array.Copy(pubBytes, 0, scriptSig, 1 + sigWithHash.Length + 1, pubBytes.Length)

                for j in 0 .. inputs.Count - 1 do
                    inputs.[j].ScriptSig <- originalScriptSigs.[j]
                inputs.[i].ScriptSig <- scriptSig
        )

    static member SelectInputs(utxos: List<UTXO>, targetAmountRtm: double, feeRateSatByte: uint32) : List<UTXO> * double =
        let targetSat = uint64 (Math.Round(targetAmountRtm * 100000000.0))
        let mutable accumulated = 0UL
        let selected = new List<UTXO>()
        let numOutputs = 2u

        let mutable finished = false
        for utxo in utxos do
            if not finished then
                selected.Add(utxo)
                accumulated <- accumulated + uint64 (Math.Round(utxo.amount * 100000000.0))
                
                let size = 148u * uint32 selected.Count + 34u * numOutputs + 10u
                let fee = uint64 size * uint64 feeRateSatByte
                
                if accumulated >= targetSat + fee then
                    finished <- true
                    
        if accumulated < targetSat then raise (new InvalidOperationException("Insufficient funds"))
        
        let size = 148u * uint32 selected.Count + 34u * numOutputs + 10u
        let finalFee = uint64 size * uint64 feeRateSatByte
        (selected, double finalFee)
